whats-that-typeclass-alternative/article.md

```diff
@@ -0,0 +1,414 @@
+                        (Nothing, rest') -> (Nothing, rest')
+```
+
+I won't go into the details of `Applicative` here, as there are [excellent other
+articles](https://serokell.io/blog/whats-that-typeclass-applicative) on the idea
+and use of applicative functors. But with that out of the way, we're finally able
+to define the `Alternative` typeclass for our parser.
+
+```haskell
+instance Alternative (Parser i) where
+    empty = Parser (\is -> (Nothing, is))
+    -- with TupleSections could be written as: Parser (Nothing,)
+
+    (<|>) :: Parser i a -> Parser i a -> Parser i a
+    (Parser pa) <|> (Parser pb) = Parser $ \is ->
+      let (aMa, restA) = pa is
+          (aMb, restB) = pb is
+      in case (aMa, aMb) of
+           (Just a, _)       -> (Just a, restA)
+           (Nothing, Just b) -> (Just b, restB)
+           _                 -> (Nothing, is)
+```
+
+But why should we actually want an `Alternative` instance for the parser?
+First of all, because of the possibility to compose multiple `Parser i a` with
+the `<|>` operator.
+
+Another reason is two methods of `Alternative` that we haven't
+discussed so far, which we get for free:
+
+* `some :: f a -> f [a]` which means: _one or more_
+* `many :: f a -> f [a]` which means: _zero of more_
+
+In our case, `some (parseLT 10)` will create a new parser of type
+`parseLT :: Int -> Parser Int [Int]` that expects at least one integer
+that is smaller 10 and then will consume all following numbers which are
+smaller 10 as well.
+
+In case of `many (parseLT 10)`, the type signature looks the same as in the
+`some` case, but it does not fail if the first element in the input isn't a
+number smaller 10. In that case, it will just return the empty list `[]`.
+
+So lets try out these methods:
+
+```none
+λ> evalParser (some $ parseLT 10) [2,3,10,20]
+Right [2,3]
+
+λ> evalParser (many $ parseLT 10) [2,3,10,20]
+Right [2,3]
+
+λ> evalParser (some $ parseLT 10) [10,2,3,20]
+Left "parser failed on [10,2,3,20]"
+
+λ> evalParser (many $ parseLT 10) [10,2,3,20]
+Right []
+```
+
+`some` and `many` are a very common and useful pattern in parser combinators.
+
+If you want to dive deeper into the topic of parsers in Haskell, take a look
+[at this post](https://serokell.io/blog/parser-combinators-in-haskell).
+
+## Conclusion and outlook
+
+* `Alternative` is a useful instance to implement for your applicative functor
+  if it has a semantic of _try this or alternatively that_.
+* `Alternative` can be seen as a monoid on applicative functors.
+* If you implement a parser (combinator), you almost always want to implement
+  an `Alternative` instance for it.
+
+Another very closely related typeclass is [`MonadPlus m`](https://wiki.haskell.org/MonadPlus)
+which has the same semantics as `Alternative` but for types which also implement `Monad`. 
+So it can be seen as a "monoid on monads". It comes with a default implementation
+based on `Alternative`, so you don't have to actually implement any methods:
```

Comment:
Could you explain when one would choose `Alternative` over `MonadPlus` or vice-versa?

Reply:
Added a paragraph about the "more powerfull"  monadic context by the example of `mfilter`.